llvm/lib/Transforms/Utils/LowerWGLocalMemoryCall.cpp

```diff
@@ -0,0 +1,114 @@
+//===-- LowerWGLocalMemoryCall.cpp - SYCL kernel local memory allocation --===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass replaces calls to __sycl_allocateLocalMemory(Size, Alignment)
+// function with allocation of memory in local address space at the kernel
+// scope.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LowerWGLocalMemoryCall.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "lowerWGLocalMemoryCall"
+
+static constexpr char SYCL_ALLOCLOCALMEM_CALL[] = "__sycl_allocateLocalMemory";
+static constexpr char LOCALMEMORY_GV_PREF[] = "WGLocalMem";
+
+namespace {
+class LowerWGLocalMemoryCallLegacy : public ModulePass {
+public:
+  static char ID;
+
+  LowerWGLocalMemoryCallLegacy() : ModulePass(ID) {
+    initializeLowerWGLocalMemoryCallLegacyPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager DummyMAM;
+    auto PA = Impl.run(M, DummyMAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  LowerWGLocalMemoryCallPass Impl;
+};
+} // namespace
+
+char LowerWGLocalMemoryCallLegacy::ID = 0;
+INITIALIZE_PASS(LowerWGLocalMemoryCallLegacy, "sycllowerwglocalmemorycall",
+                "Replace __sycl_allocateLocalMemory with allocation of memory "
+                "in local address space",
+                false, false)
+
+ModulePass *llvm::createLowerWGLocalMemoryCallPass() {
+  return new LowerWGLocalMemoryCallLegacy();
+}
+
+static bool lowerAllocaLocalMem(Module &M) {
+  SmallVector<CallInst *, 8> ToReplace;
+  for (auto &F : M) {
+    for (auto &I : instructions(F)) {
+      auto *CI = dyn_cast<CallInst>(&I);
+      Function *Callee = nullptr;
+      if (!CI || !(Callee = CI->getCalledFunction()))
+        continue;
+      StringRef Name = Callee->getName();
+      if (Name == SYCL_ALLOCLOCALMEM_CALL)
+        ToReplace.push_back(CI);
+    }
```

Comment:
It would be useful to add an assertion to verify that intrinsic allocates memory only in a kernel scope.

Reply:
+1.
Also, can this occur inside if? Inside a loop? If not - assertions could be added too.

Reply:
@kbobrovs Sorry, I didn't get this comment. Could you please clarify it?

Reply:
@mlychkov, is it legal for __sycl_allocateLocalMemory to happen:
- inside a loop
- inside a block not dominating exit
?

Reply:
@Pennycook Could you please help to answer this question?

Reply:
No, it's not allowed, but LLVM pass is not the best place to diagnose these cases. Error checking should be done in the front-end compiler.

> It would be useful to add an assertion to verify that intrinsic allocates memory only in a kernel scope.

It turns out that even check like this is difficult to implement in LLVM pass with the current design. We would like to allow `__sycl_allocateLocalMemory` invocations in lambda call operator's scope, which is technically not SPIR-V kernel scope.

Reply:
> No, it's not allowed

Right, that's what I expected. Checking validity should somehow be addressed. Plus one more case - non-convergent control flow.
So it should be checked that `__sycl_allocateLocalMemory` (or its source form - `group_local_memory`) does not occur:
- in a function (other than user lambda/functor)
- in a loop
- in a non-convergent control flow
Can we please add a TODO and file an issue?

I think it is OK to occur in convergent if. Some execution paths will just not use the global created for the allocation.